File: PADISSOPOC/src/Domain/Identity/UsernameRules.cs
using System.Text.RegularExpressions;

namespace Padi.Services.Authentication.Domain.Identity;

/// <summary>
/// The constraint Cognito enforces on the <c>Username</c> request parameter of SignUp,
/// ForgotPassword, ConfirmSignUp and the admin operations: 1–128 characters matching
/// <c>[\p{L}\p{M}\p{S}\p{N}\p{P}]+</c>. Whitespace is absent from that set, so a name
/// containing a space is rejected outright.
///
/// This has to be checked by hand because Cognito does not apply it where the value is
/// actually written. <c>preferred_username</c> is set through an attribute update, and
/// attribute values are capped at 2048 characters with no pattern at all — so Cognito
/// accepts "john smith" as an alias and then refuses it everywhere the user would type it.
/// The pool cannot help either: schema constraints support only min and max length.
/// </summary>
public static partial class UsernameRules
{
    public const int MaxLength = 128;

    /// <summary>
    /// Mirrors the Cognito Username pattern exactly — do not widen.
    ///
    /// Note that .NET applies these categories per UTF-16 code unit, so an astral-plane
    /// character such as an emoji is seen as a surrogate pair and rejected, even though
    /// the same pattern accepts it in an engine that matches by code point. The browser
    /// validator rejects astral characters explicitly so the two agree.
    /// </summary>
    [GeneratedRegex(@"^[\p{L}\p{M}\p{S}\p{N}\p{P}]+$")]
    private static partial Regex Allowed();

    /// <summary>Returns null when valid, otherwise a message naming the problem.</summary>
    public static string? Validate(string? candidate)
    {
        if (string.IsNullOrWhiteSpace(candidate))
        {
            return "Username is required.";
        }

        if (candidate != candidate.Trim())
        {
            return "Username cannot start or end with a space.";
        }

        if (candidate.Length > MaxLength)
        {
            return $"Username cannot be longer than {MaxLength} characters.";
        }

        if (candidate.Any(char.IsWhiteSpace))
        {
            return "Username cannot contain spaces.";
        }

        return Allowed().IsMatch(candidate)
            ? null
            : "Username contains a character Cognito does not accept.";
    }

    public static bool IsValid(string? candidate) => Validate(candidate) is null;
}

File: PADISSOPOC/web/src/username-rules.ts
/**
 * Mirrors src/Domain/Identity/UsernameRules.cs.
 *
 * Cognito constrains the `Username` request parameter to 1–128 characters matching
 * `[\p{L}\p{M}\p{S}\p{N}\p{P}]+` — a set that excludes whitespace entirely. It does *not*
 * apply that constraint where `preferred_username` is written: attribute values allow up to
 * 2048 characters with no pattern, so Cognito will happily store "john smith" as an alias
 * and then reject it on sign-in and password reset. Validating here is what keeps the two
 * consistent; the PostConfirmation trigger enforces the same rule server-side.
 */
export const MAX_USERNAME_LENGTH = 128;

/** The Cognito Username pattern, verbatim. Do not widen. */
const ALLOWED = /^[\p{L}\p{M}\p{S}\p{N}\p{P}]+$/u;

/** Returns null when valid, otherwise a message naming the problem. */
export function validateUsername(candidate: string): string | null {
  if (!candidate) {
    return 'Username is required.';
  }

  if (candidate !== candidate.trim()) {
    return 'Username cannot start or end with a space.';
  }

  if (candidate.length > MAX_USERNAME_LENGTH) {
    return `Username cannot be longer than ${MAX_USERNAME_LENGTH} characters.`;
  }

  if (/\s/u.test(candidate)) {
    return 'Username cannot contain spaces.';
  }

  // .NET matches \p{S} against UTF-16 code units, so an astral-plane character such as an
  // emoji lands in the surrogate category and fails the server-side check — while this
  // regex, with the /u flag, matches by code point and would accept it. Rejecting here
  // keeps the two in agreement, erring toward the stricter side: the cost is choosing a
  // different name, not a confirmation that throws after the account already exists.
  if (/[\u{10000}-\u{10FFFF}]/u.test(candidate)) {
    return 'Username cannot contain emoji or other extended characters.';
  }

  return ALLOWED.test(candidate)
    ? null
    : 'Username contains a character Cognito does not accept.';
}

export const USERNAME_RULES =
  `Up to ${MAX_USERNAME_LENGTH} characters, no spaces. Letters, digits and punctuation are allowed.`;
